lib/router.js

```diff
@@ -155,534 +27,351 @@ function compileURL(options) {
-    var _path = route.spec.path;
-    var _url = _path.replace(/\/:([A-Za-z0-9_]+)(\([^\\]+?\))?/g, pathItem);
-    var items = Object.keys(query || {}).map(queryItem);
-    var queryString = items.length > 0 ? '?' + items.join('&') : '';
-    return _url + queryString;
+    // Decorate req
+    req.route = route;
+
+    return route.chain.run.bind(route.chain);
 };
 
 /**
  * Adds a route.
  *
  * @public
+ * @memberof Router
+ * @instance
  * @function mount
- * @param    {Object} options - an options object
+ * @param    {Object} opts - an options object
+ * @param    {String} opts.name - name
+ * @param    {String} opts.method - method
+ * @param    {String} opts.path - path can be any String
+ * @param    {Function[]} handlers - handlers
  * @returns  {String} returns the route name if creation is successful.
+ * @fires ...String#mount
  */
-Router.prototype.mount = function mount(options) {
-    assert.object(options, 'options');
-    assert.string(options.method, 'options.method');
-    assert.string(options.name, 'options.name');
-
-    var exists;
-    var name = options.name;
-    var route;
-    var routes = this.routes[options.method];
+Router.prototype.mount = function mount(opts, handlers) {
     var self = this;
-    var type = options.contentType || self.contentType;
-    var versions = options.versions || options.version || self.versions;
-
-    if (type) {
-        if (!Array.isArray(type)) {
-            type = [type];
-        }
-        type
-            .filter(function filter(t) {
-                return t;
-            })
-            .sort()
-            .join();
-    }
 
-    if (versions) {
-        if (!Array.isArray(versions)) {
-            versions = [versions];
-        }
-        versions.sort();
-    }
+    assert.object(opts, 'opts');
+    assert.string(opts.method, 'opts.method');
+    assert.arrayOfFunc(handlers, 'handlers');
+    assert.optionalString(opts.name, 'opts.name');
 
-    exists = routes.some(function some(r) {
-        return r.name === name;
+    var chain = new Chain({
+        onceNext: self.onceNext,
+        strictNext: self.strictNext
     });
 
-    if (exists) {
-        return false;
-    }
-
-    route = {
-        name: name,
-        method: options.method,
-        path: compileURL({
-            url: options.path || options.url,
-            flags: options.flags,
-            urlParamPattern: options.urlParamPattern,
-            strict: self.strict
-        }),
-        spec: options,
-        types: type,
-        versions: versions
+    // Route
+    var route = {
+        name: self._getRouteName(opts.name, opts.method, opts.path),
+        method: opts.method,
+        path: opts.path,
+        spec: opts,
+        chain: chain
     };
-    routes.push(route);
 
-    if (!this.reverse[route.path.source]) {
-        this.reverse[route.path.source] = [];
-    }
+    handlers.forEach(function forEach(handler) {
+        // Assign name to anonymus functions
+        handler._name =
```

Comment:
Not sure I love that we're mutating user parameters - but wouldn't consider it blocking. Maybe worth adding a TODO for later (or filing issue).